File: Challenge 3/c3_easy.py
#Daily Programmer Challenge 3 : Easy
#@author:   Baron Daugherty
#@date:     2015-06-14
#@desc:     This program allows the user to encode some text via a Caesar Cipher
#           of a user-defined shift. The shift can be + (to the left) or - (to
#           the right) but doesn't handle shifts greater than 25 spaces (no
#           "wrap around" on shifting)

#main method drives the program... gather input and call other methods
def main():
    #gather the shift number and text to encode
    shift = int(input("Number of spaces to shift? "))
    text = input("Text to encode: ").lower()

    #get the cipher
    cipher = generate_cipher(shift)

    #output the encoded text
    print(encode(text, cipher))

#generate cipher returns a list, index 0 is the standard english 26-letter
#alphabet, index 1 is the shifted cipher alphabet
def generate_cipher(shift):
    #cretae a list and add the english alphabet to it
    alphabet = []
    alphabet.append("abcdefghijklmnopqrstuvwxyz")

    #split the alphabet into 3 parts (the middle part is the character we split
    #on and subsequently the first letter of the new alphabet
    shifted = alphabet[0].partition(alphabet[0][shift])

    #turn the tuple alphabet into one string and add it to the alphabet list
    shifted = ''.join((shifted[1], shifted[2], shifted[0]))
    alphabet.append(shifted)

    #return the list
    return alphabet

#encode turns the plaintext into ciphertext
def encode(text, cipher):
    #a list for our encoded text
    encoded = []

    #for all alpha characters, add their cipher equivalent to the list
    for letter in text:
        if letter.isalpha():
            encoded.append(cipher[1][cipher[0].index(letter)])
        else:
            encoded.append(letter)

    #return the ciphered text as one string
    return ''.join(encoded)
            
#start the program
main()

File: Challenge 3/c3_easy.pl
#!/usr/bin/perl
#Daily Programmer Challenge 3 - Easy
#@author: Baron Daugherty
#@desc: Implementation of a ROT13 cipher

use warnings;
use strict;

my $text;
#hash table maps individual characters to their corresponding character 13 places up the alphabet
#we could do fancy chr() and ord() magic, but this table allows us to just be explicit
my %cipher = ('a'=>'m', 'b'=>'n', 'c'=>'o', 'd'=>'p', 'e'=>'q', 'f'=>'r', 'g'=>'s',
              'h'=>'t', 'i'=>'u', 'j'=>'v', 'k'=>'w', 'l'=>'x', 'm'=>'y', 'n'=>'z',
              'o'=>'a', 'p'=>'b', 'q'=>'c', 'r'=>'d', 's'=>'e', 't'=>'f', 'u'=>'g',
              'v'=>'h', 'w'=>'i', 'x'=>'j', 'y'=>'k', 'z'=>'l');

#prompt for and collect input, ignore case
print("Enter some text to encrypt: ");
chomp($text = lc(<>));

#loop through the characters in the input and output as appropriate
foreach my $letter (split(//,$text)){
    #if the character is not in the hash table, print out the character as is
    if (not exists $cipher{$letter}) {
        print($letter);
    }
    else{
        #otherwise, print the appropriate value for the key
        print($cipher{$letter});
    }
}  #END loop